### PR TITLE
checkpoint also QAT for resuming vol.2

### LINKED ISSUE #6457: Can't load_from_checkpoint with QuantizationAwareTraining callback during training
for PL v1.2.0

```py
LightningModel.load_from_checkpoint(str(ckpt_path))
```
gives
```
  File "//checkpoint/utils.py", line 11, in load_pl_model
    pl_model: LightningModel = LightningModel.load_from_checkpoint(str(ckpt_path))
  File "/.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 156, in load_from_checkpoint
    model = cls._load_model_state(checkpoint, strict=strict, **kwargs)
  File "//.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 204, in _load_model_state
    model.load_state_dict(checkpoint['state_dict'], strict=strict)
  File "//.venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1223, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightningModel:
Unexpected key(s) in state_dict: "quant.activation_post_process.fake_quant_enabled", "quant.activation_post_process.observer_enabled", "quant.activation_post_process.scale", "quant.activation_post_process.zero_point", "quant.activation_post_process.activation_post_process.eps", "quant.activation_post_process.activation_post_process.min_val", "quant.activation_post_process.activation_post_process.max_val", "model.encoder._bn0.activation_post_process.fake_quant_enabled", "model.encoder._bn0.activation_post_process.observer_enabled", "model.encoder._bn0.activation_post_process.scale", "model.encoder._bn0.activation_post_process.zero_point", .............
```

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Fixes #6457
Closes #8485

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @tchaton @rohitgr7 @akihironitta @borda

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1085,11 +1085,12 @@ def tune(
     def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None) -> None:
         # restore modules after setup
         self.checkpoint_connector.resume_start(checkpoint_path)
-        self.checkpoint_connector.restore_model()
-        self.checkpoint_connector.restore_datamodule()
+        # FIXME
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:1088]: @awaelchli @carmocca so the suggestion here was to add a protected method on `checkpoint_connector` for callback restore which would be used only by this QAT callback, right?

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -217,6 +218,33 @@ def restore_precision_plugin_state(self) -> None:
         ):
             prec_plugin.load_state_dict(self._loaded_checkpoint["native_amp_scaling_state"])
 
+    def restore_quantization_callbacks(self) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:221]: I added this method

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -1140,6 +1140,7 @@ def tune(
     def _restore_modules_and_callbacks(self, checkpoint_path: Optional[_PATH] = None) -> None:
         # restore modules after setup
         self._checkpoint_connector.resume_start(checkpoint_path)
+        self._checkpoint_connector.restore_quantization_callbacks()
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:1143]: I added this call

File: pytorch_lightning/callbacks/quantization.py [hunk 4]
```diff
@@ -311,3 +334,19 @@ def on_predict_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule
...
+        arg_names = ("qconfig", "observer_type", "collect_quantization", "modules_to_fuse", "input_compatible")
+        attribs = {n: getattr(self, f"_{n}") for n in arg_names}
+        return attribs
+
+    def load_before_model(self, pl_module: "pl.LightningModule", state_dict: Dict[str, Any]) -> None:
+        """Special hook that gets called by the CheckpointConnector *before* the model gets loaded.
+
+        This hook replaces the :meth:`on_load_checkpoint` and :meth:`load_state_dict` callback methods which get called
+        after the model has already loaded the weights. For quantization, we need to convert the model first before
+        that happens, assuming the previous training used quantization.
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/quantization.py:348]: Check that my explanation is correct. 
Also, note in the last sentence the assumption we make.

**Review by carmocca — APPROVED** [review]
@Borda does this need a CHANGELOG update?

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/callbacks/quantization.py [hunk 5]
```diff
@@ -178,7 +182,7 @@ def __init__(
             )
         self._collect_quantization = collect_quantization
 
-        self.modules_to_fuse = modules_to_fuse
+        self._modules_to_fuse = modules_to_fuse
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/quantization.py:185]: Guess it's covered by:

>     .. warning:: ``QuantizationAwareTraining`` is in beta and subject to change.

**Pull request merged**